Backports for Julia 1.11.8

gbaraldi opened the pull request:
Backported PRs:
- [x] #57523 <!-- Remove usages of weak symbols -->
- [x] #58127 <!-- [DOC] Update installation docs: /downloads/ => /install/ -->
- [x] #58202 <!-- [release-1.11] malloc: use jl_get_current_task to fix null check -->
- [x] #58554 <!-- remove workaround for controlling terminal behavior in repl_cmd -->
- [x] #58584 <!-- Make `Ptr` values static-show w/ type-information -->
- [x] #59062 <!-- remove a testset from MMAP that might cause CI to now fail on Windows -->
- [x] #59300 <!-- Update the developer docs to reflect the use of JuliaSyntax.jl -->
- [x] #57604 <!-- `@nospecialize` for `string_index_err` -->
- [x] #59329 <!-- aotcompile: destroy LLVM context after serializing combined module -->
- [x] #59418 <!-- Fix startup when history file is bad -->
- [x] #56890 <!-- Enable getting non-boxed LLVM type from Julia Type -->
- [x] #59559 <!-- codegen: mark write barrier field load as volatile -->
- [x] #59572 <!-- Only apply Base.Sort.SubArrayOptimization when iszero(v.offset1) -->

Need manual backport:
- [ ] #56329 <!-- loading: clean up more concurrency issues -->
- [ ] #56956 <!-- Add "mea culpa" to foreign module assignment error. -->
- [ ] #57035 <!-- linux: workaround to avoid deadlock inside dl_iterate_phdr in glibc -->
- [ ] #57089 <!-- Block thread from receiving profile signal with stackwalk lock -->
- [ ] #57249 <!-- restore non-freebsd-unix fix for profiling -->
- [ ] #58011 <!-- Remove try-finally scope from `@time_imports` `@trace_compile` `@trace_dispatch` -->
- [ ] #58062 <!-- remove unnecessary edge from `exp_impl` to `pow` -->
- [ ] #58157 <!-- add showing a string to REPL precompile workload -->
- [ ] #58209 <!-- Specialize `one` for the `SizedArray` test helper -->
- [ ] #58356 <!-- codegen: remove readonly from abstract type calling convention -->
- [ ] #58415 <!-- [REPL] more reliable extension loading -->
- [ ] #58510 <!-- Don't filter `Core` methods from newly-inferred list -->
- [ ] #58110 <!-- relax dispatch for the `IteratorSize` method for `Generator` -->
- [ ] #58965 <!-- Fix `hygienic-scope`s in inner macro expansions -->
- [ ] #58971 <!-- Fix alignment of failed precompile jobs on CI -->
- [ ] #59066 <!-- build: Also pass -fno-strict-aliasing for C++ -->
- [ ] #59428 <!-- Correctly set the variant bits of uuid1 -->

Contains multiple commits, manual intervention needed:
- [ ] #55877 <!-- fix FileWatching designs and add workaround for a stat bug on Apple -->
- [ ] #56755 <!-- docs: fix scope type of a `struct` to hard -->
- [ ] #57809 <!-- Fix fptrunc Float64 -> Float16 rounding through Float32 -->
- [ ] #57398 <!-- Make remaining float intrinsics require float arguments -->
- [ ] #56351 <!-- Fix `--project=@script` when outside script directory -->
- [ ] #57129 <!-- clarify that time_ns is monotonic -->
- [ ] #58134 <!-- Note annotated string API is experimental in Julia 1.11 in HISTORY.md -->
- [ ] #58401 <!-- check that hashing of types does not foreigncall (`jl_type_hash` is concrete evaluated) -->
- [ ] #58435 <!-- Fix layout flags for types that have oddly sized primitive type fields -->
- [ ] #58483 <!-- Fix tbaa usage when storing into heap allocated immutable structs -->
- [ ] #58512 <!-- Make more types jl_static_show readably -->
- [ ] #58012 <!-- Re-enable tab completion of kwargs for large method tables -->
- [ ] #58683 <!-- Add 0 predecessor to entry basic block and handle it in inlining -->
- [ ] #59112 <!-- Add builtin function name to add methods error -->
- [ ] #56823 <!-- Make version of opaque closure constructor in world -->
- [ ] #59467 <!-- `CoreLogging`: prevent some `Annotated*`-related instability -->

Non-merged PRs with backport label:
- [ ] #59450 <!-- Propagate Addrspaces: fix lift of memset -->
- [ ] #58848 <!-- Set array size only when safe to do so -->
- [ ] #55958 <!-- also redirect JL_STDERR etc. when redirecting to devnull -->
- [ ] #55956 <!-- Make threadcall gc safe -->
- [ ] #55534 <!-- Set stdlib sources as read-only during installation -->
- [ ] #55499 <!-- propagate the terminal's `displaysize` to the `IOContext` used by the REPL -->
- [ ] #55458 <!-- Allow for generically extracting unannotated string -->
- [ ] #55457 <!-- Make AnnotateChar equality consider annotations -->
- [ ] #55220 <!-- `isfile_casesensitive` fixes on Windows -->
- [ ] #53957 <!-- tweak how filtering is done for what packages should be precompiled -->
- [ ] #51479 <!-- prevent code loading from lookin in the versioned environment when building Julia -->
- [ ] #50813 <!-- More doctests for Sockets and capitalization fix -->
- [ ] #50157 <!-- improve docs for `@inbounds` and `Base.@propagate_inbounds` -->

vtjnash committed:
[REPL] more reliable extension loading (#58415)

A weird edge case of loading, if REPL is loaded explicitly and does not
come from a require_stdlib call, it should not be getting REPLExt from a
require_stdlib call either. This is a convoluted bit of hackery specific
to work around problems with the way Pkg's REPLExt is designed to mutate
the REPL in unsafe ways. No other stdlib should ever want to access an
extension, particularly of a different module, as that is a private API
violation on multiple counts, so this only needs to be made to work
specifically for that REPL-Pkg scenario, even if it looks seemingly more
general.

Refs https://github.com/JuliaLang/julia/issues/58373

Reproducer:
```
JULIA_DEPOT_PATH=tmpdir/.julia ./julia --hist=no -qie 'using REPL'
] status
```

vtjnash committed:
remove workaround for controlling terminal behavior in repl_cmd (#58554)

This was originally added as a workaround for the behavior of sh to try
to become the terminal process group leader, but no longer relevant
since #25006 removed that flag again:
```
julia> run(detach(`bash -i -c "sleep 0"`))
bash: cannot set terminal process group (-1): Inappropriate ioctl for device
bash: no job control in this shell
Process(`bash -i -c 'sleep 0'`, ProcessExited(0))
```

Long explanation: Julia recieves SIGTTOU when a process like "sh -i -c"
exits (at least for bash and zsh, but not dash, ksh, or csh),
since "sh -i" sometimes takes over the controlling terminal, causing
julia to stop once the bash process exits. This can be quite annoying,
but julia goes through some pains (in libuv) to ensure it doesn't steal
the controlling terminal from the parent, and apparently bash is not so
careful about it. However, since PR #25006, julia hasn't needed this
workaround for this issue, so we can now follow the intended behavior
when the child is in the same session group and steals the controlling
terminal from the parent. Even if such behavior seems odd, this seems
to be the intended behavior per posix design implementation that it
gets SIGTTOU when julia later tries to change mode (from cooked -> raw
after printing the prompt):
http://curiousthing.org/sigttin-sigttou-deep-dive-linux.

For some background on why this is a useful signal and behavior and
should not be just blocked, see nodejs/node#35536.

According to glibc, there's a half page of code for how to correctly
implement a REPL mode change call:
https://www.gnu.org/software/libc/manual/html_node/Initializing-the-Shell.html

```
$ ./julia -q
shell> bash -i -c "sleep 1"

[1]+  Stopped                 ./julia

julia> run(`zsh -i -c "sleep 0"`)
Process(`zsh -i -c 'sleep 0'`, ProcessExited(0))

julia>
[1]+  Stopped                 ./julia
```

(cherry picked from commit 0cbe4669079249e3d8ff48688b45ed3b39db20d5)

gbaraldi committed:
Make late gc lower handle insertelement of alloca use. (#58637)

This was in DAECompiler.jl code found by @serenity4. He also mentioned
that writing up how one might go and fix a bug like this so i'll give a
quick writeup (this was a very simple bug so it might not be too
interesting)

The original crash which looked something like
>   %19 = alloca [10 x i64], align 8
  %155 = insertelement <4 x ptr> poison, ptr %19, i32 0
Unexpected instruction
> [898844] signal 6 (-6): Aborted
in expression starting at
/home/gbaraldi/DAECompiler.jl/test/reflection.jl:28
pthread_kill at /lib/x86_64-linux-gnu/libc.so.6 (unknown line)
gsignal at /lib/x86_64-linux-gnu/libc.so.6 (unknown line)
abort at /lib/x86_64-linux-gnu/libc.so.6 (unknown line)
RecursivelyVisit<llvm::IntrinsicInst,
LateLowerGCFrame::PlaceRootsAndUpdateCalls(llvm::ArrayRef<int>, int,
State&, std::map<llvm::Value*, std::pair<int, int>
>)::<lambda(llvm::AllocaInst*&)>::<lambda(llvm::Use&)> > at
/home/gbaraldi/julia4/src/llvm-late-gc-lowering.cpp:803
operator() at /home/gbaraldi/julia4/src/llvm-late-gc-lowering.cpp:2560
[inlined]
PlaceRootsAndUpdateCalls at
/home/gbaraldi/julia4/src/llvm-late-gc-lowering.cpp:2576
runOnFunction at
/home/gbaraldi/julia4/src/llvm-late-gc-lowering.cpp:2638
run at /home/gbaraldi/julia4/src/llvm-late-gc-lowering.cpp:2675
run at
/home/gbaraldi/julia4/usr/include/llvm/IR/PassManagerInternal.h:91

which means it was crashing inside of late-gc-lowering, so the first
thing I did was ran julia and the same test with LLVM_ASSERTIONS=1 and
FORCE_ASSERTIONS=1 to see if LLVM complained about a malformed module,
and both were fine. Next step was trying to get the failing code out for
inspection.
Easiest way is to do `export
JULIA_LLVM_ARGS="--print-before=LateLowerGCFrame --print-module-scope"`
and pipe the output to a file.
The file is huge, but since it's a crash in LLVM we know that the last
thing is what we want, and that gave me the IR I wanted.
To verify that this is failing I did `make -C src install-analysis-deps`
to install the LLVM machinery (opt...). That gets put in the `tools`
directory of a julia build. Then I checked if this crashed outside of
julia by doing
`./opt -load-pass-plugin=../lib/libjulia-codegen.dylib
--passes=LateLowerGCFrame -S test.ll -o tmp3.ll `. This is run from
inside the tools dir so your paths might vary (the -S is so LLVM doesn't
generate bitcode) and my code did crash, however it was over 500 lines
of IR which makes it harder to debug and to write a test.

Next step then is to minimize the crash by doing
[`llvm-reduce`](https://llvm.org/docs/CommandGuide/llvm-reduce.html)
over it (it's basically creduce but optimized for LLVM IR) which gave me
a 2 line reproducer (in this case apparently just having the
insertelement was enough for the pass to fail). One thing to be wary is
that llvm-reduce will usually make very weird code, so it might be
useful to modify the code slightly so it doesn't look odd (it will have
unreachable basic-blocks and such).
After the cleanup fixing the bug here wasn't interesting but this
doesn't apply generally. And also always transform your reduced IR into
a test to put in llvmpasses.

(cherry picked from commit 906d3482d318868e7d3eef6b1db97ef5bedd6f82)

ajwheeler committed:
Update the developer docs to reflect the use of JuliaSyntax.jl (#59300)

[The "parsing" section of the "eval" dev docs
page](https://docs.julialang.org/en/v1/devdocs/eval/#dev-parsing) was
not updated when JuliaSyntax.jl was adopted in 1.10. This updates the
text to reflect that it is the default parser and briefly mentions how
to switch back.

(cherry picked from commit ec272746d5e31d8c49ce079b6dc7177f1104eb14)

nsajko committed:
`@nospecialize` for `string_index_err` (#57604)

The fields of `StringIndexError` are abstractly typed, so there's no
reason to specialize on a concrete type. The change seems like
it could prevent some invalidation on loading user code.

---------

Co-authored-by: Shuhei Kadowaki <40514306+aviatesk@users.noreply.github.com>
(cherry picked from commit bc33a3e354ac2de1ae1138b55947665278e7e2e8)

xal-0 committed:
aotcompile: destroy LLVM context after serializing combined module (#59329)

We already save some memory here by deleting the `jl_native_code_desc_t`
after
we're done serializing the combined module, but some data in the
module's
`LLVM::Context` lives on until the end of the scope in
`jl_dump_native_impl`.
Once we're done with the module, move the lock and `ThreadSafeContext`
so the
reference count drops to zero.

A crude measurement shows that when compiling the Base sysimage, about 3
GiB is
in use. Deleting the `jl_native_code_desc_t` (as before) saves about 600
MiB,
and cleaning up the context saves an additional ~500 MiB.

(cherry picked from commit ceeb661e2eaa14d3e2e174095c5747f7a7124725)

IanButterworth committed:
Fix startup when history file is bad (#59418)

(cherry picked from commit 6c02a2142c2ea88f08bf4876feca139cb21fc232)

wsmoses committed:
Enable getting non-boxed LLVM type from Julia Type (#56890)

The current method :jl_type_to_llvm takes a pointer-to-bool parameter
isboxed, which if non-null, is set to true if the Julia Type requires
boxing. However, one may want to know the julia type without boxing and
there is no mechanism for doing so. Now there is `julia_struct_to_llvm`.

(cherry picked from commit 85f1b8c9d5defbe7f0368cc5f5a3049bdec1a7ce)

gbaraldi committed:
[backports-release-1.11] Don't forget to decay value on struct initialization (#59511)

Fixes https://github.com/JuliaLang/julia/issues/59510

DilumAluthge committed:
[`backports-release-1.11`] Backport #59259 ("Fix compile time regression for null ptr comparisons") to Julia 1.11.x (#59446)

Backports #59259 to 1.11 (cherry picked from commit
920df7abc9b5bf47d0589d9768b41e41421ad683).

For more details, see
https://github.com/JuliaLang/julia/pull/59445#issuecomment-3239579009.

Targets `backports-release-1.11` (#59336).

---------

Co-authored-by: Gabriel Baraldi <baraldigabriel@gmail.com>

xal-0 committed:
[backports-release-1.11] Backport #57523: Remove usages of weak symbols  (#59470)

> Inspired by a bug I hit on MinGW recently with weak symbol resolution.
I'm not sure why we started using these in
70000ac7c3d5d5f21e42555cdf99e699a246f8ec, since I believe we should be
able to lookup these symbols just fine in our `jl_exe_handle` as long as
it was linked / compiled with `-rdynamic`
>
> Migrating away from weak symbols seems a good idea, given their uneven
implementation across platforms.

(cherry picked from commit 4a6ada6eb36eaeb9bde9b94c01bcf6f0e3ce7ae6)

@gbaraldi asked if this could be backported separately from #59227.

Co-authored-by: Cody Tapscott <84105208+topolarity@users.noreply.github.com>
Co-authored-by: Dilum Aluthge <dilum@aluthge.com>

gbaraldi committed:
Remove bad test

vtjnash committed:
codegen: mark write barrier field load as volatile (#59559)

This is as an alternative to changing all allocation functions to
mutating all memory and returning an aliasing pointer. This operates
usually late in the pipeline, so either should not make much difference,
but I think it should suffice to mark this volatile to prevent any
de-duplication of this load, and this should also be most conservative
for GC but more liberal for other optimizations.

Fixes #59547

Produced with dubious help by Claude.

(cherry picked from commit 218f691d16edbda33647d2b938b882a7ac08f057)

wsmoses committed:
Fix mingw windows build

vchuravy committed:
[release-1.11] Fix missing CodeInstance owner lookup in _jl_invoke (#59558)

Fixed on main as 92fc06fd83a, but not fixed in earlier Julia versions

DilumAluthge commented:
Gabriel has an LLVM bump cooking currently. Once that's been pushed, we'll work to get CI green, and once that's done, I'll start the PkgEval process.

LilithHafner committed:
Only apply Base.Sort.SubArrayOptimization when iszero(v.offset1) (#59572)

Fixes #59569

Thanks @N5N3 for finding the bug!

For 1.13, we could re-enable this optimization and propagate the
information to the sub-alg or have the sub-alg's embedded indices be
relative to `kw.lo` rather than absolute. For backports, this minimal
change is more appropriate.

(cherry picked from commit 067b0133a8ec3f5c81d89525cd6cb4f25df6cb9c)

gbaraldi committed:
Bump LLVM version

gbaraldi committed:
Remove test

DilumAluthge committed:
CI 1.11: Use the 1.11-specific Buildkite branch (#59760)

nanosoldier commented:
The package evaluation job [you requested](https://github.com/JuliaLang/julia/pull/59521#issuecomment-3456590725) has completed - possible new issues were detected.
The [**full report**](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/58327cc_vs_f2b3dbd/report.html) is available.

<details><summary>Report summary</summary>

## ❗ Packages that crashed

**3 packages crashed only on the current version.**

 - The process was aborted: 2 packages
 - A segmentation fault happened: 1 packages

81 packages crashed on the previous version too.

## ✖ Packages that failed

**36 packages failed only on the current version.**

 - Package has test failures: 5 packages
 - Package tests unexpectedly errored: 2 packages
 - Networking-related issues were detected: 1 packages
 - Tests became inactive: 3 packages
 - Test duration exceeded the time limit: 25 packages

3160 packages failed on the previous version too.

## ✔ Packages that passed tests

**37 packages passed tests only on the current version.**

 - Other: 37 packages

6473 packages passed tests on the previous version too.

## ➖ Packages that were skipped altogether

**1 packages were skipped only on the current version.**

 - Package could not be installed: 1 packages

1315 packages were skipped on the previous version too.



</details>

DilumAluthge commented:
@nanosoldier `runtests(["FloatTracker", "JuliaInterpreter", "Keccak", "ColPack", "UCX", "SparseMatrixColorings", "RemoteFiles", "Andes", "LazyAlgebra", "PointNeighbors", "MRICoilSensitivities", "FactorRotations", "FindMinimaxPolynomial", "BaytesOptim", "ElementaryFluxModes", "AdvancedHMC", "MCMCTempering", "MIRT", "JetPack", "BetaML", "JudiLing", "MixedModels", "MaxEntropyGraphs", "ILMPostProcessing", "Plots", "LongwaveModePropagator", "Hedgehog", "SIAN", "Yunir", "LibTrixi", "ControlBarrierFunctions", "GeneralizedSasakiNakamura", "UnfoldDecode", "GasChromatographySimulator", "Chamber", "REPL", "GIRFReco"])`

nanosoldier commented:
The package evaluation job [you requested](https://github.com/JuliaLang/julia/pull/59521#issuecomment-3474977488) has completed - possible new issues were detected.
The [**full report**](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/58327cc_vs_f2b3dbd/report.html) is available.

<details><summary>Report summary</summary>

## ✖ Packages that failed

**2 packages failed only on the current version.**

 - Package has test failures: 1 packages
 - Package tests unexpectedly errored: 1 packages

3 packages failed on the previous version too.

## ✔ Packages that passed tests

32 packages passed tests on the previous version too.



</details>

nanosoldier commented:
The package evaluation job [you requested](https://github.com/JuliaLang/julia/pull/59521#issuecomment-3475315715) has completed - possible new issues were detected.
The [**full report**](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/58327cc_vs_f2b3dbd/report.html) is available.

<details><summary>Report summary</summary>

## ✖ Packages that failed

**2 packages failed only on the current version.**

 - Package has test failures: 1 packages
 - Package tests unexpectedly errored: 1 packages




</details>

DilumAluthge commented:
Okay, so the current status of what's failing in Nanosoldier:
1. `REPL`: No idea why it's failing in Nanosoldier, but it's passing in the Buildkite CI here, so I think we can ignore the Nanosoldier failure.
2. JuliaInterpreter.jl: It looks like the same `check_builtins` failure that Kristoffer previously said we can ignore.

So I'm interpreting this as a clean Nanosoldier run on this PR.

And the Buildkite CI is all green on this PR.

And when I look at the Buildkite logs, I see that the external stdlibs were correctly NOT skipped. Example log ([link](https://buildkite.com/julialang/julia-release-1-dot-11/builds/548#0199bbcf-84bf-4500-a132-2bcaf3c5a15d)):

```
Branch is: backports-release-1.11
Pipeline is: julia-release-1-dot-11
On important branch 'backports-release-1.11' or pipeline 'julia-release-1-dot-11': running all external stdlib tests
```

And also:

```
JULIA_CMD_FOR_TESTS is:    julia-58327cce5e/bin/julia .buildkite/utilities/timeout.jl julia-58327cce5e/bin/julia
JULIA_NUM_THREADS is:      1
NCORES_FOR_TESTS is:       16
OPENBLAS_NUM_THREADS is:   16
TESTS is:                  all --ci
USE_RR is:
JL_TERM_TIMEOUT is:        110m
Core dump pattern:         %e-pid%p-sig%s-ts%t.core
Core dump size limit:      unlimited
```

DilumAluthge commented:
Therefore I think this PR is good to merge.

DilumAluthge commented:
Kristoffer and I are going to move forward with 1.11.8, in the interest of unblocking the use cases described above.

@PallHaraldsson For the PRs you mentioned, can you make sure that those PRs have the `backport 1.12` label on them (if they are appropriate for backporting)? Then we can try to get those PRs into the next 1.12.x release, at least.
